Voter dApp & CLI: Hardcode whitelist of supported price requests and filter in voter clients

nicholaspai opened the pull request:
Signed-off-by: Nick Pai <npai.nyc@gmail.com>

<!--
  Thanks for submitting a pull request!
  We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory.
-->

<!--
  Title
  Please include a concise title that briefly describes the change.
  Titles should be prefixed with the area of code that the change affects and they should be in the present simple tense. Examples:
  
  dvm: adds a new function to compute voting rewards offchain
  monitor bot: fixes broken link in liquidation log
  voter dapp: adds countdown timer component to the header
-->


**Motivation**

#1819 


**Summary**

Whitelist is imported into voter clients: CLI and Voter dApp (both active and resolved requests)


**Details**

This is obviously going to require ongoing maintenance work if this is the adopted solution.


**Issue(s)**

<!-- This PR must fix or refer to one or more issues. Please list them here. -->
Fixes #1819

nicholaspai commented on common/WhitelistedPriceRequests.js:
```diff
@@ -0,0 +1,18 @@
+// Whitelisted price requests that will display on voter clients.
+const REQUEST_WHITELIST = {
+  USDETH: ["1596666977"],
```
Not sure why quotes were stripped away here but it was due to linting

nicholaspai commented:
> High-level question, why do a whitelist rather than a blacklist?

This is a great question. It depends do we want to whitelist or blacklist by default. I was thinking that if a series of spam requests come, then whitelisting by default would prevent them from showing up on the dApp. The tradeoff though is that we need to deploy an update for every valid request. @chrismaree @daywiss @kendricktan @hlambur this is not an easy question WDYT?

> Also, is there a toggle to allow voters to see the "spam" requests if they want?

No, should we? Does that defeat the purpose?

mrice32 commented:
> > Also, is there a toggle to allow voters to see the "spam" requests if they want?
> 
> No, should we? Does that defeat the purpose?

We discussed this in the meeting a bit. The idea is that it keeps usability high, but also means that we're not _preventing_ voters from seeing or voting on requests if they disagree with our curation. It's kinda a middle ground to avoid being accused of centralized censorship. Thoughts?

chrismaree commented:
> > > Also, is there a toggle to allow voters to see the "spam" requests if they want?
> > 
> > 
> > No, should we? Does that defeat the purpose?
> 
> We discussed this in the meeting a bit. The idea is that it keeps usability high, but also means that we're not _preventing_ voters from seeing or voting on requests if they disagree with our curation. It's kinda a middle ground to avoid being accused of centralized censorship. Thoughts?

This looks like a good starting point. I have three points to raise:
1) This should be a blacklist, not a whitelist. We should assume that in the default case things will be fine and don't need us to remove stuff. Only if there is spam should it be filtered out. We don't want to have to update this file _all the time_ when there are normal votes but rather only update it when we absolutely have to.
2) As @mrice32 pointed out, we don't actually want to prevent users from seeing this info, just to make it clear that it's spam via a "show spam votes" button of some kind. Basically this could sit at the top (or bottom) of the table and give the user the option to include, if they click it, votes that we've flagged as spam. By default, they are hidden. If the user chooses to view these then it should potentially highlight which votes are spam.
3) I think we should consider not storing this blacklist in GitHub for the simple reason that it is slower to update than we might want. Ideally, we should store it in a Gist or some other form that we can more quickly iterate and respond to if need be.

daywiss commented:
I would also lean towards blacklist for the same reason chris mentions. If we anticipate that we see more spam than legit disputes, then we should prefer whitelist, currently i think we are optimistic that wont happen.

on hiding/warning:  I personally think its better to hide with an opt in to see. But if that is significant work then we can add that as a seperate PR. 

I'm ok with this being git repo or gist, but probably need to poll whatever the source is to make sure its getting the latest without refreshing page.

mrice32 commented:
@chrismaree agree on points 1 & 2 -- I think this should be a blacklist and it should be toggle-able with a button.

> 3. I think we should consider not storing this blacklist in GitHub for the simple reason that it is slower to update than we might want. Ideally, we should store it in a Gist or some other form that we can more quickly iterate and respond to if need be that does not require us going through CI in the case we need to respond to a spam situation.

We can't create organization gists, and the permissioning isn't super clear to me. Thoughts on creating just a separate repo that anyone in the org can push to directly? Or maybe that just requires "review" from one other person (but no CI)? There may be other ways to manage this data, too. Definitely open to other ideas if anyone has them.

nicholaspai commented:
@chrismaree @daywiss @mrice32 I added 2 things and left the gist vs repo JSON decision on hold for now:
- Blacklist instead of whitelist requests
- Add spam toggle. Kind of ugly right now, but need one switch for each component since there are two separate components to render conditionally.

![Screen Shot 2020-08-07 at 10 59 26](https://user-images.githubusercontent.com/9457025/89659048-51ad5300-d89d-11ea-835e-9819378c0d92.png)

chrismaree commented:
> @chrismaree @daywiss @mrice32 I added 2 things and left the gist vs repo JSON decision on hold for now:
> 
> * Blacklist instead of whitelist requests
> * Add spam toggle. Kind of ugly right now, but need one switch for each component since there are two separate components to render conditionally.
> 
> ![Screen Shot 2020-08-07 at 10 59 26](https://user-images.githubusercontent.com/9457025/89659048-51ad5300-d89d-11ea-835e-9819378c0d92.png)

this looks pretty good! one small refinement that'll make it cleaner is to hide the toggle switch if we have not added any spam requests. So the toggle _only_ shows if it actually does something (i.e there is spam to filter).

nicholaspai commented:
> > @chrismaree @daywiss @mrice32 I added 2 things and left the gist vs repo JSON decision on hold for now:
> > 
> > * Blacklist instead of whitelist requests
> > * Add spam toggle. Kind of ugly right now, but need one switch for each component since there are two separate components to render conditionally.
> > 
> > ![Screen Shot 2020-08-07 at 10 59 26](https://user-images.githubusercontent.com/9457025/89659048-51ad5300-d89d-11ea-835e-9819378c0d92.png)
> 
> this looks pretty good! one small refinement that'll make it cleaner is to hide the toggle switch if we have not added any spam requests. So the toggle _only_ shows if it actually does something (i.e there is spam to filter).

Great idea, I commented out the example identifier in REQUEST_BLACKLIST and changed the render logic for the spam Switch to:

```
{Object.keys(REQUEST_BLACKLIST).length > 0 && (
        <FormGroup>
          <FormControlLabel
            control={
              <Switch size="small" checked={showSpamRequests} onChange={() => setShowSpamRequests(!showSpamRequests)} />
            }
            label="Show spam price requests"
          />
        </FormGroup>
      )}
```

mrice32 commented on voter-dapp/src/ActiveRequests.js:
```diff
@@ -603,6 +623,16 @@ function ActiveRequests({ votingAccount, votingGateway }) {
           </DialogContentText>
         </DialogContent>
       </Dialog>
+      {Object.keys(REQUEST_BLACKLIST).length > 0 && (
```
Let's say that there is a price request that's been blacklisted in the past, but it's the request is longer active (for whatever reason). Would this toggle still show despite the fact that it won't do anything?

Not sure if this is a case that matters. It may be that these requests will always stay active after being blacklisted, so if there are any blacklisted, they will be in this table.

nicholaspai replied:
> Would this toggle still show despite the fact that it won't do anything?

yes this toggle would show. Perhaps I should check that there are no current pending requests on the blacklist

nicholaspai replied:
@mrice32 I changed the `Object.keys(REQUEST_BLACKLIST).length > 0` condition to: 

```
      {allPendingRequests.length >= pendingRequests.length && (<Component/>)
```

This will show the switch IFF there are any requests from `allPendingRequests` that have been filtered out.

mrice32 replied:
Nice and simple -- I like it

chrismaree commented on common/BlacklistedPriceRequests.js:
```diff
@@ -0,0 +1,8 @@
+// Whitelisted price requests that will display on voter clients.
```
```suggestion
// Blacklisted price requests that will display on voter clients.
```

chrismaree commented on voter-dapp/src/ResolvedRequests.js:
```diff
@@ -162,6 +184,17 @@ function ResolvedRequests({ votingAccount }) {
           </List>
         )}
       </Dialog>
+      {/* Only render this spam filter switch if some resolved spam requests have been filtered out */}
```
+1

chrismaree commented on common/BlacklistedPriceRequests.js:
```diff
@@ -0,0 +1,8 @@
+// Whitelisted price requests that will display on voter clients.
+const REQUEST_BLACKLIST = {
+  SOME_IDENTIFIER: ["1596666977"]
```
this is just a dummy identifier, right?

nicholaspai replied:
Yep just wanted to show the format. There is no harm in leaving this uncommented as `ActiveRequests` and `ResolvedRequests` only care if a pending or resolved request matches with something on this list
